[onechat] Add ES|QL Tool

dennis-tismenko opened the pull request:
## Summary

This PR introduces the ability to create ES|QL tools from the Kibana UI. The form is built with React Hook Form and validation is implemented using Zod and a custom resolver.

### Demoes

__Happy Path: Successful submission__

https://github.com/user-attachments/assets/11e92547-4628-4933-b5fd-c51d85094daf

__Validation__

https://github.com/user-attachments/assets/53232c2c-4bbd-4bcf-b61a-49236a5a9910

__Submission Error__

https://github.com/user-attachments/assets/d1e4dc73-6d01-47a8-bdec-f804eb0f6b23


### Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [x] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [x] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

dennis-tismenko committed:
Implement add ES|QL tool functionality

This change includes the full stack implementation for adding ES|QL
tools, including the necessary frontend validation.

dennis-tismenko committed:
Refactor: Cleanup 1 character variable names

dennis-tismenko committed:
Refactor: Separate into components, hooks, and utils

dennis-tismenko committed:
Refactor: Fix compilation errors, fix missing tags in payload, submit state not resetting

dennis-tismenko commented on x-pack/platform/plugins/shared/onechat/public/application/components/tools/esql/form/hooks/use_esql_editor_params.ts:
```diff
@@ -0,0 +1,181 @@
+    listeners.push(modelCreationListener, commandListener, completionProvider);
+
+    return () => {
+      listeners.forEach((listener) => listener.dispose());
```
Not an issue

dennis-tismenko commented on x-pack/platform/plugins/shared/onechat/public/application/components/tools/esql/form/components/esql_editor_field.tsx:
```diff
@@ -0,0 +1,205 @@
+                    triggerEsqlParamWarnings();
+                  }
+                }}
+                onTextLangQuerySubmit={promisify(noop)} // Required prop, but we don't need it
```
Definitely some room for improvement in the component interface (eg. lack of `onError` callback)

kibanamachine committed:
[CI] Auto-commit changed files from 'ts-node .buildkite/pipeline-resource-definitions/scripts/fix-location-collection.ts'

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/319490)
* Commit: c927e5a3bc5df255101dbceaf8424940bee31e81

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/e98699000285a403566162782d57acaa6f18c238) | [after](https://github.com/elastic/kibana/commit/c927e5a3bc5df255101dbceaf8424940bee31e81) | diff |
| --- | --- | --- | --- |
| `onechat` | 138 | 371 | +233 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/e98699000285a403566162782d57acaa6f18c238) | [after](https://github.com/elastic/kibana/commit/c927e5a3bc5df255101dbceaf8424940bee31e81) | diff |
| --- | --- | --- | --- |
| `@kbn/onechat-common` | 144 | 146 | +2 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/e98699000285a403566162782d57acaa6f18c238) | [after](https://github.com/elastic/kibana/commit/c927e5a3bc5df255101dbceaf8424940bee31e81) | diff |
| --- | --- | --- | --- |
| `onechat` | 70.9KB | 688.3KB | :warning: +617.3KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/e98699000285a403566162782d57acaa6f18c238) | [after](https://github.com/elastic/kibana/commit/c927e5a3bc5df255101dbceaf8424940bee31e81) | diff |
| --- | --- | --- | --- |
| `onechat` | 11.1KB | 11.4KB | +310.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/e98699000285a403566162782d57acaa6f18c238) | [after](https://github.com/elastic/kibana/commit/c927e5a3bc5df255101dbceaf8424940bee31e81) | diff |
| --- | --- | --- | --- |
| `@kbn/onechat-common` | 219 | 221 | +2 |


</details>


cc @dennis-tismenko

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/319490","number":319490,"commit":"c927e5a3bc5df255101dbceaf8424940bee31e81"}],"number":319490}
buildkite-pr-comment-->

jedrazb reviewed (commented):
UI feels great! Great stuff. Found couple of issues, described them in comments

meghanmurphy1 commented on x-pack/platform/plugins/shared/onechat/public/application/components/tools/esql/esql_tools.tsx:
```diff
@@ -0,0 +1,167 @@
+          })}
+        </EuiText>
+        <EuiHorizontalRule margin="xs" />
+        <EuiInMemoryTable
```
Why are we using EuiInMemoryTable ? Seems like it automatically handles pagination and sorting but we may want to have more control over that.

dennis-tismenko replied:
In short: to get basic search and pagination working. Agree it's not scalable long-term, but the goal here was to get a UX win over a full-fledged API based solution

pgayvallet replied:
Agreed, InMemoryTable is the pragmatic solution until we implement server-side paging (and the volume of tools we have should be very fine for a long time)

dennis-tismenko commented on x-pack/platform/plugins/shared/onechat/public/application/components/tools/esql/form/components/esql_editor_field.tsx:
```diff
@@ -0,0 +1,205 @@
+    }
+  }, []);
+
+  const { triggerEsqlParamWarnings, triggerEsqlParamFieldsValidation } = useEsqlParamsValidation();
```
Awesome catch! Figured out that it was a bug in the validation resolver, will fix

afharo reviewed (approved):
`kibana.jsonc` changes LGTM

dennis-tismenko commented on x-pack/platform/plugins/shared/onechat/public/application/components/tools/esql/esql_tools.tsx:
```diff
@@ -0,0 +1,167 @@
+          error={errorMessage}
+          search={search}
+          pagination={{
+            pageIndex: 0,
```
Yep, good catch, I did not test pagination at all — will fix in this PR if it's as simple as just tracking in React state.

dennis-tismenko commented on x-pack/platform/plugins/shared/onechat/public/application/utils/extract_esql_params.ts:
```diff
@@ -0,0 +1,30 @@
+ * @param esql The ES|QL query string.
+ * @returns An array of RegExpMatchArray objects for each parameter found.
+ */
+export const extractEsqlParamMatches = (esql: string): RegExpMatchArray[] =>
```
Looks to be the same, will refactor to use this util

meghanmurphy1 commented on x-pack/platform/plugins/shared/onechat/public/application/components/tools/esql/form/types/esql_tool_form_types.ts:
```diff
@@ -0,0 +1,26 @@
+  warning?: string;
+};
+
+export interface OnechatEsqlToolFormData {
```
Should we be using `EsqlToolDefintion` from `@kbn/onechat-common` for consistency here?

dennis-tismenko committed:
Fix: tool table pagination, validation bug, scrollIntoView, i18n

dennis-tismenko committed:
Fix: param validation, filtering unused params

dennis-tismenko committed:
Fix: mobile

dennis-tismenko committed:
Fix: submission error handling

dennis-tismenko commented on x-pack/platform/plugins/shared/onechat/public/application/components/tools/esql/create_esql_tool_flyout.tsx:
```diff
@@ -0,0 +1,202 @@
+  const handleSave = useCallback(
+    async (data: OnechatEsqlToolFormData) => {
+      try {
+        const tool = await createTool(transformEsqlFormData(data));
```
Yep, good catch. Updated to use the mutation's callbacks.. There's also a bug in older versions of RHF that don't properly set `isSubmitting` when the promise rejects, so I'll use the mutation's `isLoading` instead — other form functionality such as validation appears intact.

dennis-tismenko commented on x-pack/platform/plugins/shared/onechat/public/application/components/tools/esql/create_esql_tool_flyout.tsx:
```diff
@@ -0,0 +1,202 @@
+  const { isSubmitting, errors, isSubmitSuccessful } = formState;
+  const { euiTheme } = useEuiTheme();
+
+  useEffect(() => {
+    if (isSubmitSuccessful) {
+      reset(defaultValues);
+    }
+  }, [isSubmitSuccessful, reset]);
```
As you mentioned elsewhere, RHF treating it as a successful form submission is a side effect of swallowing the error, but if submission is actually successful, this `useEffect` is still the recommended approach for resetting the form state as opposed to in `onSuccess` (RHF quirk, see discussion [here](https://github.com/react-hook-form/react-hook-form/issues/8741))

kibanamachine committed:
[CI] Auto-commit changed files from 'ts-node .buildkite/pipeline-resource-definitions/scripts/fix-location-collection.ts'

dennis-tismenko commented on x-pack/platform/plugins/shared/onechat/public/application/components/tools/esql/create_esql_tool_flyout.tsx:
```diff
@@ -0,0 +1,202 @@
+  onError?: (error: Error) => void;
+}
+
+const defaultValues: OnechatEsqlToolFormData = {
```
Yeah, really like your implementation, I'll refactor into a hook when in a follow-up PR when introducing editing

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/320245)
* Commit: 8245680eb218060081220d3b6db3d371e2bb417c

### Failed CI Steps
* [Jest Tests #18](https://buildkite.com/elastic/kibana-pull-request/builds/320245#01981559-e448-4806-96a6-3b1fba6e671a)
* [Jest Tests #18](https://buildkite.com/elastic/kibana-pull-request/builds/320245#0198157f-455f-4374-b5c9-75519d68a7f7)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/320245#01981559-e448-4806-96a6-3b1fba6e671a) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/320245/jobs/01981559-e448-4806-96a6-3b1fba6e671a/artifacts/0198157f-2fe5-404f-ab47-7a8bed23201c) Jest Tests #<span></span>18 / transformEsqlFormData should transform ES|QL form data to a create tool payload
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/320245#0198157f-455f-4374-b5c9-75519d68a7f7) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/320245/jobs/0198157f-455f-4374-b5c9-75519d68a7f7/artifacts/01981583-99b4-48a5-9998-4da8477b3f74) Jest Tests #<span></span>18 / transformEsqlFormData should transform ES|QL form data to a create tool payload

### History
* :green_heart: [Build #319490](https://buildkite.com/elastic/kibana-pull-request/builds/319490) succeeded c927e5a3bc5df255101dbceaf8424940bee31e81

cc @dennis-tismenko

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/320245","number":320245,"commit":"8245680eb218060081220d3b6db3d371e2bb417c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/319490","number":319490,"commit":"c927e5a3bc5df255101dbceaf8424940bee31e81"}],"number":320245}
buildkite-pr-comment-->

jedrazb reviewed (commented):
Tested locally and feels great! 👍 

the only nit is that I have esql error immediately (empty input) even before touching the input ... but it's a nit

<img width="817" height="899" alt="Image" src="https://github.com/user-attachments/assets/10cc0af5-d908-4bf8-8cc0-d02599acabd9" />

My other comment is about validation, and using common logic as much as possible, see my comment about matching params

dennis-tismenko commented on x-pack/platform/plugins/shared/onechat/public/application/components/tools/esql/create_esql_tool_flyout.tsx:
```diff
@@ -0,0 +1,202 @@
+  css`
+    position: relative;
+
+    .euiFlyoutBody__overflow {
+      transform: none; /* Remove the transform to avoid creating a new stacking context */
+      mask-image: none; /* Remove the mask-image to avoid cutting off ES|QL suggestions */
+    }
+
+    /* Manually re-add the same fade effect to the flyout body */
+    &::before {
+      content: '';
+      position: absolute;
+      top: 0;
+      left: 0;
+      right: 0;
+      height: ${euiTheme.size.s};
+      background: linear-gradient(to bottom, ${euiTheme.colors.plainLight}, transparent);
+      pointer-events: none;
+    }
```
Right yeah, but the issue primarily lies in the `ESQLLangEditor` component and how it display the suggestions rather than the flyout itself — although I'm not sure if there's even a way to display suggestions properly over the `mask-image`, so maybe that is a `EuiFlyout` limitation

dennis-tismenko committed:
fix: initial validation for esql, failing tests

jedrazb reviewed (approved):
LGTM! Great stuff, tested locally.

pgayvallet reviewed (commented):
LGTM
